tools/btmidi-server.c

```diff
@@ -0,0 +1,927 @@
+	if (!mgmt_if) {
+		fprintf(stderr, "Failed to access management interface\n");
+		return -1;
+	}
+
+	data = bt_ad_new();
+	if (!data) {
+		fprintf(stderr, "Error creating adverting data\n");
+		return -1;
+	}
+
+	if (!bt_ad_add_service_uuid(data, &uuid)) {
+		fprintf(stderr, "Error adding service UUID\n");
+		return -1;
+	}
+
+	adv_data = bt_ad_generate(data, &adv_data_len);
+	if (!adv_data) {
+		fprintf(stderr, "Error generating advertising data\n");
+		return -1;
+	}
+
+	param_len = sizeof(struct mgmt_cp_add_advertising) + adv_data_len;
+	cp = malloc0(param_len);
+	if (!cp) {
+		fprintf(stderr, "Error allocating memory to advertising params\n");
+		return -1;
+	}
+
+	cp->flags = htobl(MGMT_ADV_FLAG_CONNECTABLE | MGMT_ADV_FLAG_DISCOV);
+	cp->instance = 0;
+	cp->adv_data_len = adv_data_len;
+	memcpy(cp->data, adv_data, adv_data_len);
+
+	free(adv_data);
+
+	if (!mgmt_send(mgmt_if, MGMT_OP_ADD_ADVERTISING, dev_id, param_len,
+				cp, add_adv_callback, NULL, NULL)) {
+		fprintf(stderr, "Failed to add Advertising Data");
+		free(cp);
+		return -1;
+	}
+
+	free(cp);
+	bt_ad_unref(data);
+	mgmt_unref(mgmt_if);
+
+	return 0;
+}
+
+static int l2cap_le_att_listen_and_accept(bdaddr_t *src, int sec,
+							uint8_t src_type)
+{
+	int sk, nsk;
+	struct sockaddr_l2 srcaddr, addr;
+	socklen_t optlen;
+	struct bt_security btsec;
+	char ba[18];
+	int dev_id;
+	
+	sk = socket(PF_BLUETOOTH, SOCK_SEQPACKET, BTPROTO_L2CAP);
+	if (sk < 0) {
+		perror("Failed to create L2CAP socket");
+		return -1;
+	}
+
+	dev_id = hci_get_route(src);
+	if (dev_id < 0) {
+		errno = ENODEV;
+		fprintf(stderr, "Device ID not found\n");
+		goto fail;
+	}
+
+	if (advertise(dev_id) != 0) {
+		goto fail;
+	}
+
+	/* Set up source address */
+	memset(&srcaddr, 0, sizeof(srcaddr));
+	srcaddr.l2_family = AF_BLUETOOTH;
+	srcaddr.l2_cid = htobs(ATT_CID);
+	srcaddr.l2_bdaddr_type = src_type;
+	bacpy(&srcaddr.l2_bdaddr, src);
+
+	if (bind(sk, (struct sockaddr *) &srcaddr, sizeof(srcaddr)) < 0) {
+		perror("Failed to bind L2CAP socket");
+		goto fail;
+	}
+
+	/* Set the security level */
+	memset(&btsec, 0, sizeof(btsec));
+	btsec.level = sec;
+	if (setsockopt(sk, SOL_BLUETOOTH, BT_SECURITY, &btsec,
+							sizeof(btsec)) != 0) {
+		fprintf(stderr, "Failed to set L2CAP security level\n");
+		goto fail;
+	}
+
+	if (listen(sk, 10) < 0) {
+		perror("Listening on socket failed");
+		goto fail;
+	}
+
+	if (verbose) {
+		printf("Started listening on ATT channel. Waiting for connections\n");
+	}
+
+	memset(&addr, 0, sizeof(addr));
+	optlen = sizeof(addr);
+	nsk = accept(sk, (struct sockaddr *) &addr, &optlen);
+	if (nsk < 0) {
+		perror("Accept failed");
+		goto fail;
+	}
+
+	if (verbose) {
+		ba2str(&addr.l2_bdaddr, ba);
+		printf("Connect from %s\n", ba);
+	}
+	
+	close(sk);
+
+	return nsk;
+
+fail:
+	close(sk);
+	return -1;
+}
+
+static struct server *server_create(int fd, uint16_t mtu)
+{
+	struct server *server;
+	size_t name_len = strlen(gattName);
+	int err;
+
+	server = new0(struct server, 1);
```

Comment:
You will never free this object if this function succeed. Also, your `goto` approach is not safe, you are potentially freeing stuff that might not be allocated. You should free objects in order as they get created.
And `goto` is not legacy. ;)